### ISSUE TITLE
distributed testing

### ISSUE THREAD

**Issue by LiangTinting** [ruinmessi/ASFF#20]
Hi there, thanks for sharing your code!
I have tested your pre-trained model "YOLOv3 800+ ASFF*" with distributed testing(on 4 TITAN XPs) and single GPU testing(on one TITAN XP) on COCO minival. I noticed that except the inference time gap between them, the single GPU testing's performance is 1% mAP lower than the distributed testing results. Could you please explain that for me?

The single GPU result:                        67.32ms, 42.6mAP
Distributed test on 4 GPUs result:     88.41ms, 43.6mAP

Thank you so much for your time!

**Comment by ruinmessi** [comment]
Hi, thanks for reproducing the results. Since the inference time is also different, I guess you may test the model with different input sizes. Did you use '-s' to set the same test input size?

**Comment by LiangTinting** [comment]
Yes, the input size is 800

**Comment by LiangTinting** [comment]
Also, I test on 4 GPUs without distributed testing, the result is the same as single GPU, 42.6mAP, the inference time is the same as distributed test on 4 GPUs.

**Comment by ruinmessi** [comment]
ok, I will try to reproduce it and figure out what happened.

**Comment by ruinmessi** [comment]
@LiangTinting hi, I have reproduced this issue when I test the model on a single GPU. It is quite weird. I reviewed my code and find nothing.....I guess maybe there is something wrong with apex.parallel.DistributedDataParallel, and I'll have a closer look if I have time at the weekend. Thanks!!

**Comment by CuongNguyen218** [comment]
I hope you can find out what happen.

**Comment by ruinmessi** [comment]
@LiangTinting @CuongNguyen218  Hey, guys! I finally the bug in my evaluation code.

**Comment by LiangTinting** [comment]
Great! What was that?




------------------&nbsp;原始邮件&nbsp;------------------
发件人: "ruinmessi"<notifications@github.com&gt;; 
发送时间: 2019年12月1日(星期天) 中午12:15
收件人: "ruinmessi/ASFF"<ASFF@noreply.github.com&gt;; 
抄送: "梁婷婷"<775083792@qq.com&gt;; "Mention"<mention@noreply.github.com&gt;; 
主题: Re: [ruinmessi/ASFF] distributed testing (#20)




@LiangTinting @CuongNguyen218  Hey, guys! I finally the bug in my evaluation code.

**Comment by ruinmessi** [comment]
@LiangTinting @CuongNguyen218  Hey guys! I finally find the bug in my evaluation code. Actually it has several issues:
For the inference time gap, the dist testing actually accumulates more total time and divide fewer samples, that's why the single gpu testing is significantly faster.
For mAP gap, I made two mistakes. One is the NMS setting, I took a look at my original code and find that actually I use 0.001 conf threshhold and 0.65 nms thresh when I evaluation on test-dev, which can get better performance and also longer nms time. Another bug is " cocoEval.params.imgIds = ids" in cocoapi_evaluator. With this code, we only evaluate the results on the main gpu. So 42.6 is the right results with the old nms setting (0.01 conf, 0.6 iou).
Now, I add the eval.py file for pure evaluation and accumulate the inference time correctly in cocoapi_evaluator.py. For both single and dist testing, the mAP is 43.5 on COCO minival.
